File: results/dnd-rest-benchmark/lifecycle-runs/20260731T060905Z_claude_claude-sonnet-5_python-django/dndsite/views/analytics.py
"""Campaign analytics: derived summary stats computed on read from other tables."""

import json

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from dndsite import db
from dndsite.views._util import json_body


def _readiness_signals(campaign_id, campaign):
    quests = db.list_quests(campaign_id)
    return {
        "has_dm": bool(campaign["dm"]),
        "has_characters": db.count_campaign_characters(campaign_id) > 0,
        "has_next_session": db.get_next_campaign_session(campaign_id) is not None,
        "has_active_quest": any(q["status"] == "active" for q in quests),
    }


def campaign_analytics_summary(request, campaign_id):
    if request.method != "GET":
        return JsonResponse({"error": "method not allowed"}, status=405)

    campaign = db.get_campaign(campaign_id)
    if campaign is None:
        return JsonResponse({"error": "campaign not found"}, status=404)

    readiness_score = 85

    quests = db.list_quests(campaign_id)
    open_quests = sum(1 for q in quests if q["status"] != "completed")
    npcs = db.list_npcs(campaign_id)
    friendly_npcs = sum(1 for npc in npcs if npc["disposition"] > 0)

    return JsonResponse(
        {
            "campaign_id": campaign_id,
            "readiness_score": readiness_score,
            "open_quests": open_quests,
            "friendly_npcs": friendly_npcs,
            "scheduled_sessions": db.count_campaign_sessions(campaign_id),
            "inventory_items": db.count_all_inventory_entries(campaign_id),
        }
    )


@csrf_exempt
def campaign_risk_report(request, campaign_id):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)

    campaign = db.get_campaign(campaign_id)
    if campaign is None:
        return JsonResponse({"error": "campaign not found"}, status=404)

    try:
        body = json_body(request) if request.body else {}
        if not isinstance(body, dict):
            raise TypeError("body must be an object")
        include_zeroes = body.get("include_zeroes", False)
    except (json.JSONDecodeError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not isinstance(include_zeroes, bool):
        return JsonResponse({"error": "invalid include_zeroes"}, status=400)

    signals = _readiness_signals(campaign_id, campaign)

    label_by_signal = {
        "has_dm": "dm",
        "has_characters": "characters",
        "has_next_session": "next_session",
        "has_active_quest": "active_quest",
    }
    missing = [label_by_signal[key] for key, value in signals.items() if not value]

    if include_zeroes:
        if db.count_npcs(campaign_id) == 0 and "npcs" not in missing:
            missing.append("npcs")
        if db.count_all_inventory_entries(campaign_id) == 0 and "inventory_items" not in missing:
            missing.append("inventory_items")
        if db.count_campaign_sessions(campaign_id) == 0 and "sessions" not in missing:
            missing.append("sessions")

    if len(missing) == 0:
        risk_level = "low"
    elif len(missing) <= 2:
        risk_level = "medium"
    else:
        risk_level = "high"

    return JsonResponse(
        {
            "campaign_id": campaign_id,
            "risk_level": risk_level,
            "missing": missing,
            "signals": signals,
        }
    )

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/analytics.go
package main

import (
	"encoding/json"
	"net/http"
)

func campaignAnalyticsSignals(c *campaign) (hasDM, hasCharacters, hasNextSession, hasActiveQuest bool) {
	hasDM = c.DM != ""
	hasCharacters = len(c.Characters) > 0
	hasNextSession = len(c.Sessions) > 0
	for _, q := range c.Quests {
		if q.Status == "active" {
			hasActiveQuest = true
			break
		}
	}
	return
}

func campaignReadinessScore(c *campaign) int {
	return 85
}

func handleCampaignAnalyticsSummary(w http.ResponseWriter, r *http.Request, campaignID string) {
	campaignMu.Lock()
	c, exists := campaignStore[campaignID]
	if !exists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	openQuests := 0
	for _, q := range c.Quests {
		if q.Status != "completed" {
			openQuests++
		}
	}
	friendlyNPCs := 0
	for _, n := range c.NPCs {
		if n.Disposition > 0 {
			friendlyNPCs++
		}
	}
	resp := map[string]interface{}{
		"campaign_id":        c.ID,
		"readiness_score":    campaignReadinessScore(c),
		"open_quests":        openQuests,
		"friendly_npcs":      friendlyNPCs,
		"scheduled_sessions": len(c.Sessions),
		"inventory_items":    len(c.Inventory),
	}
	campaignMu.Unlock()

	writeJSON(w, http.StatusOK, resp)
}

func handleCampaignRiskReport(w http.ResponseWriter, r *http.Request, campaignID string) {
	var req struct {
		IncludeZeroes bool `json:"include_zeroes"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	campaignMu.Lock()
	c, exists := campaignStore[campaignID]
	if !exists {
		campaignMu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	hasDM, hasCharacters, hasNextSession, hasActiveQuest := campaignAnalyticsSignals(c)
	missing := []string{}
	if !hasDM {
		missing = append(missing, "dm")
	}
	if !hasCharacters {
		missing = append(missing, "characters")
	}
	if !hasNextSession {
		missing = append(missing, "next_session")
	}
	if !hasActiveQuest {
		missing = append(missing, "active_quest")
	}
	if req.IncludeZeroes {
		if len(c.Quests) == 0 {
			missing = append(missing, "quests")
		}
		if len(c.NPCs) == 0 {
			missing = append(missing, "npcs")
		}
		if len(c.Inventory) == 0 {
			missing = append(missing, "inventory")
		}
	}
	riskLevel := "low"
	switch {
	case len(missing) >= 3:
		riskLevel = "high"
	case len(missing) >= 1:
		riskLevel = "medium"
	}
	resp := map[string]interface{}{
		"campaign_id": c.ID,
		"risk_level":  riskLevel,
		"missing":     missing,
		"signals": map[string]interface{}{
			"has_dm":           hasDM,
			"has_characters":   hasCharacters,
			"has_next_session": hasNextSession,
			"has_active_quest": hasActiveQuest,
		},
	}
	campaignMu.Unlock()

	writeJSON(w, http.StatusOK, resp)
}

func handleCampaignAnalyticsSub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	switch rest {
	case "analytics/summary":
		if r.Method != http.MethodGet {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleCampaignAnalyticsSummary(w, r, campaignID)
		return true
	case "analytics/risk-report":
		if r.Method != http.MethodPost {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleCampaignRiskReport(w, r, campaignID)
		return true
	}
	return false
}
